File: 分割回文串.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        int n = s.size();
        vector<vector<string>> ans;
        vector<string> path;

        function<bool(string, int, int)> judge = [&](string s, int left, int right) -> bool {
            while (left < right) {
                if (s[left++] != s[right--]) {
                    return false;
                }
            }
            return true;
        };

        function<void(int)> dfs = [&](int i) {
            if (i == n) {
                ans.emplace_back(path);
                return;
            }
            for (int j = i; j < n; ++j) {
                if (judge(s, i, j)) {
                    path.emplace_back(s.substr(i, j - i + 1));
                    dfs(j + 1);
                    path.pop_back();
                }
            }
        };

        dfs(0);
        return ans;
    }
};

File: 分割回文串.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        n = len(s)
        ans, path = [], []

        def judge(s, left, right):
            while left < right:
                if s[left] != s[right]:
                    return False
                left, right = left + 1, right - 1
            return True
        
        def dfs(i):
            if i == n:
                ans.append(path.copy())
                return
            for j in range(i, n):
                if judge(s, i, j):
                    path.append(s[i:j + 1])
                    dfs(j + 1)
                    path.pop()

        dfs(0)
        return ans
